fix(emotion): fix render loop

RobinClowers opened the pull request:
This solves the render loop when loading the emotion example. As far as I can see, we only need to re-inject the styles on the client if the top level component gets unmounted (because that removes the document head, where the styles are).

My original PR had removed all of the code related to this re-injection, but we do need it for that case.

Original description below.

~However, when you trigger the catch boundary on the client, you loose all the styles, since the entire document is re-rendered. This same problem exists in the Chakra example, but you can't easily tell, since there is no link you can follow to trigger the error boundary on the client.~

I finally understand what the code I was removed was trying to do (add the styles back in after re-mounting the document), but it was broken.

~I'm not sure what the correct fix is, but I think this is better than the broken example we have now. I will try to spend some more time on this soon, to see if I can find a complete solution.~

Fixes: https://github.com/remix-run/examples/issues/1

MichaelDeBoey reviewed (commented):
I updated this in #40 to be in line with what @chaance did in #35
Since I'm not familiar with `emotion`, I'll let him have a look at this one

RobinClowers commented:
@MichaelDeBoey Thanks for the quick response! I looked at this a bit more and I think I found a more correct solution that works with top level error and catch boundaries as well.

coreyward reviewed (approved):
I have no permissions here, but I just want to confirm that this change works for me. I would love to see this applied since the example is misleading/broken as it stands!

RobinClowers commented:
Agreed, this could probably be improved further, but it works, and the example is currently broken, so it seems like a clear win. @chaance can you find time to take a look at this?

jmborden commented:
@RobinClowers Thanks for this! Funny to bump into each other again in such a random place :D

chaance commented on emotion/app/root.tsx:
```diff
@@ -50,7 +52,9 @@ const Document = withEmotionCache(
       clientStyleData.reset();
-    }, [clientStyleData, emotionCache.sheet]);
+      // We only want to do this on mount, not when styles change
+      // eslint-disable-next-line react-hooks/exhaustive-deps
```
This won't work as expected anyway since effects may run more than once (in `StrictMode` this effect will always be called twice as of React 18)

RobinClowers replied:
We could use a ref to track if we've executed the re-linking, which would solve the issue of it being called twice. If we did that, we could avoid disabling the lint rule, though I would argue this is one of the rare cases when it makes sense to disable the rule.

chaance replied:
If you go that route it wouldn't make a difference if the effect ran once or 1000 times really, so I'd opt not to disable the rule to keep examples minimal and clean (and to prevent folks from thinking we're sneakily breaking rules when we're not).

RobinClowers replied:
I'll make that change tonight, hopefully we can get this merged!

dani-mp replied:
Btw, this example still uses React 17's `hydrate` and `renderToString` if I'm not mistaken.

chaance replied:
We can bump to React 18, but I'm pretty sure it needs to use `renderToString` if styles are rendered in the head vs. adjacent to component elements. The docs suggest streaming APIs are not supported with the "advanced" approach. https://emotion.sh/docs/ssr#advanced-approach

dani-mp replied:
Yeah, there's also https://emotion.sh/docs/ssr#renderstylestonodestream, but I'm super confused.

I arrived here because I have two projects, one built with Next and the other with Remix, and both use MUI (and hence Emotion). I'm trying to upgrade both to React 18 and I'm having a hard time understanding everything.

chaance replied:
Welcome to our world! We're trying to get people out of this stuff for a reason 😅

RobinClowers commented:
@jmborden 🥂 sure thing!

RobinClowers commented:
I've pushed up the ref change so the relink won't happen twice with strictmode in dev and got rid of the lint disable. I think this should be merged, since it fixes the issue, if we want to update to react 18, that could be a separate PR.

RobinClowers commented:
@chaance anything preventing this from merging?

MichaelDeBoey commented on emotion/app/root.tsx:
```diff
@@ -35,21 +35,28 @@ const Document = withEmotionCache(
     useEffect(() => {
-      // re-link sheet container
-      emotionCache.sheet.container = document.head;
+      if (reinjectStylesRef.current) {
```
What do you think about using an early return here?

```suggestion
      if (!reinjectStylesRef.current) {
        return;
      }
```

MichaelDeBoey commented on emotion/app/root.tsx:
```diff
@@ -35,21 +35,28 @@ const Document = withEmotionCache(
+        clientStyleData.reset();
+        // ensure we only do this once per mount
+        reinjectStylesRef.current = false;
+      }
```
```suggestion
```

MichaelDeBoey commented on emotion/app/root.tsx:
```diff
@@ -35,21 +35,29 @@ const Document = withEmotionCache(
-      tags.forEach((tag) => {
-        (emotionCache.sheet as any)._insertTag(tag);
-      });
+        // re-inject tags
+        const tags = emotionCache.sheet.tags;
+        emotionCache.sheet.flush();
+        tags.forEach((tag) => {
+          (emotionCache.sheet as any)._insertTag(tag);
+        });
```
```suggestion
      // re-inject tags
      const tags = emotionCache.sheet.tags;
      emotionCache.sheet.flush();
      tags.forEach((tag) => {
        (emotionCache.sheet as any)._insertTag(tag);
      });
```

MichaelDeBoey reviewed (changes requested):
Be sure to run ESLint/Prettier on your files

RobinClowers commented:
> Be sure to run ESLint/Prettier on your files

Sorry about that, I committed your suggestions through github. I just fixed this and squashed the fix into the previous commit. Thanks for looking at this!

MichaelDeBoey reviewed (approved):
Awesome!
Thanks for taking the time to update this example @RobinClowers! 🙏

If you would like to also create a PR to update this one to React 18, we'd be happy to accept a PR for that as well
